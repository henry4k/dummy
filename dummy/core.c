#include <stdlib.h> // malloc, free
#include <string.h> // memset, strncpy
#include <stdio.h>  // dummyFormat etc
#include <stdarg.h> // dummyFormat etc
#include <stdbool.h>
#include <assert.h>

#include "core.h"


enum
{
    DUMMY_MAX_MESSAGE_LENGTH = 127,
    DUMMY_MAX_TESTS = 32,
    DUMMY_MAX_CLEANUPS = 32,
    DUMMY_INVALID_TEST_INDEX = -1,
    DUMMY_SANDBOX_SKIPPED = -1
};

typedef enum
{
    DUMMY_INITIALIZING,
    DUMMY_RUNNING,
    DUMMY_COMPLETED
} dummyStatus;

typedef enum
{
    DUMMY_TEST_UNDEFINED,
    DUMMY_TEST_STARTING,
    DUMMY_TEST_RUNNING,
    DUMMY_TEST_COMPLETED
} dummyTestStatus;

/**
 * Is called after the test has been completed -
 * regardless of whether it failed or succeeded.
 */
typedef struct
{
    dummyCleanupFunction fn;
    void* data;
} dummyCleanup;

typedef struct
{
    char name[DUMMY_MAX_MESSAGE_LENGTH];
    dummySandbox sandbox;
    dummySandboxableFunction fn;

    dummyTestStatus status;

    dummyTestResult result;
    char abortReason[DUMMY_MAX_MESSAGE_LENGTH];

    bool markedAsTodo;
    char todoReason[DUMMY_MAX_MESSAGE_LENGTH];

    dummyCleanup cleanupStack[DUMMY_MAX_CLEANUPS];
    int cleanupStackSize;
} dummyTest;

typedef struct
{
    dummyStatus status;
    const dummyReporter* reporter;

    dummyTest tests[DUMMY_MAX_TESTS];
    int testCount;
    int currentTestIndex;
} dummyContext;

typedef struct
{
    dummyAbortHandler handler;
    void* context;
} dummyAbortHandlerInfo;


static dummyContext* dummyCurrentContext = NULL;
static dummyAbortHandlerInfo AbortStack[DUMMY_MAX_SANDBOX_DEPTH];
static int AbortStackSize = 0;


bool dummyRunTest( int index );
dummyTest* dummyGetCurrentTest();

void dummyInit( const dummyReporter* reporter )
{
    assert(dummyCurrentContext == NULL);
    dummyCurrentContext = (dummyContext*)malloc(sizeof(dummyContext));
    memset(dummyCurrentContext, 0, sizeof(dummyContext));

    dummyCurrentContext->status = DUMMY_INITIALIZING;

    assert(reporter);
    assert(reporter->began);
    assert(reporter->completed);
    assert(reporter->beganTest);
    assert(reporter->completedTest);
    assert(reporter->log);
    dummyCurrentContext->reporter = reporter;

    dummyCurrentContext->currentTestIndex = DUMMY_INVALID_TEST_INDEX;
}

int dummyRunTests()
{
    dummyContext* ctx = dummyCurrentContext;
    assert(ctx);
    assert(ctx->status == DUMMY_INITIALIZING);

    ctx->reporter->began(ctx->reporter->context);

    ctx->status = DUMMY_RUNNING;
    int failedTests = 0;
    int i = 0;
    for(; i < ctx->testCount; i++)
        if(dummyRunTest(i) == false)
            failedTests++;

    ctx->status = DUMMY_COMPLETED;
    ctx->reporter->completed(ctx->reporter->context);

    assert(ctx->status == DUMMY_COMPLETED);
    free(ctx);
    dummyCurrentContext = NULL;

    return failedTests;
}

void dummyAddTest( const char* name, dummySandbox sandbox, dummySandboxableFunction fn )
{
    dummyContext* ctx = dummyCurrentContext;
    assert(ctx);
    assert(ctx->status == DUMMY_INITIALIZING);
    assert(ctx->testCount < DUMMY_MAX_TESTS);

    dummyTest* test = &ctx->tests[ctx->testCount];
    memset(test, 0, sizeof(dummyTest));

    strncpy(test->name, name, DUMMY_MAX_MESSAGE_LENGTH);
    test->sandbox = sandbox;
    test->fn = fn;

    ctx->testCount++;
}

bool dummyRunTest( int index )
{
    dummyContext* ctx = dummyCurrentContext;
    assert(ctx);
    assert(ctx->status == DUMMY_RUNNING);

    assert(index >= 0);
    assert(index < DUMMY_MAX_TESTS);
    dummyTest* test = &ctx->tests[index];

    // prepare
    ctx->currentTestIndex = index;
    test->status = DUMMY_TEST_STARTING;
    test->result = DUMMY_TEST_PASSED;
    ctx->reporter->beganTest(ctx->reporter->context);

    // run
    test->status = DUMMY_TEST_RUNNING;
    const char* abortReason = NULL;
    const int errorCode = test->sandbox(test->fn, &abortReason);
    switch(errorCode)
    {
        case DUMMY_SANDBOX_SUCCEEDED:
            test->result = DUMMY_TEST_PASSED;
            break;

        case DUMMY_SANDBOX_SKIPPED:
            test->result = DUMMY_TEST_SKIPPED;
            break;

        case DUMMY_SANDBOX_GENERIC_ERROR:
        default:
            test->result = DUMMY_TEST_FAILED;
    }
    if(abortReason)
        strncpy(test->abortReason, abortReason, DUMMY_MAX_MESSAGE_LENGTH);
    test->status = DUMMY_TEST_COMPLETED;

    for(int i = 0; i < test->cleanupStackSize; i++)
    {
        dummyCleanup* cleanup = &test->cleanupStack[i];
        cleanup->fn(cleanup->data);
    }
    test->cleanupStackSize = 0;

    // cleanup
    ctx->reporter->completedTest(ctx->reporter->context);
    ctx->currentTestIndex = DUMMY_INVALID_TEST_INDEX;
    test->status = DUMMY_TEST_UNDEFINED;

    return test->result != DUMMY_TEST_FAILED;
}

int dummyGetTestCount()
{
    const dummyContext* ctx = dummyCurrentContext;
    assert(ctx);

    return ctx->testCount;
}

int dummyGetTestNumber()
{
    const dummyContext* ctx = dummyCurrentContext;
    assert(ctx);

    assert(ctx->currentTestIndex >= 0);
    assert(ctx->currentTestIndex < DUMMY_MAX_TESTS);

    return ctx->currentTestIndex;
}

dummyTest* dummyGetCurrentTest()
{
    dummyContext* ctx = dummyCurrentContext;
    assert(ctx);

    assert(ctx->currentTestIndex >= 0);
    assert(ctx->currentTestIndex < DUMMY_MAX_TESTS);
    dummyTest* test = &ctx->tests[ctx->currentTestIndex];

    return test;
}

const char* dummyGetTestName()
{
    const dummyTest* test = dummyGetCurrentTest();
    assert(test->name);
    assert(strlen(test->name) < DUMMY_MAX_MESSAGE_LENGTH);
    return test->name;
}

dummyTestResult dummyGetTestResult()
{
    const dummyTest* test = dummyGetCurrentTest();
    assert(test->status == DUMMY_TEST_COMPLETED);
    return test->result;
}

const char* dummyGetTestAbortReason()
{
    const dummyTest* test = dummyGetCurrentTest();
    if(test->status == DUMMY_TEST_COMPLETED)
    {
        assert(test->abortReason);
        const int length = strlen(test->abortReason);
        assert(length < DUMMY_MAX_MESSAGE_LENGTH);
        if(length > 0)
            return test->abortReason;
    }

    return NULL;
}

void dummyAddCleanup( dummyCleanupFunction fn, void* data )
{
    dummyTest* test = dummyGetCurrentTest();

    assert(test->cleanupStackSize < DUMMY_MAX_CLEANUPS);
    dummyCleanup* cleanup = &test->cleanupStack[test->cleanupStackSize];

    cleanup->fn = fn;
    cleanup->data = data;

    test->cleanupStackSize++;
}

int dummyTestIsMarkedAsTodo()
{
    return dummyGetCurrentTest()->markedAsTodo;
}

const char* dummyGetTestTodoReason()
{
    const dummyTest* test = dummyGetCurrentTest();

    if(test->markedAsTodo)
    {
        assert(test->todoReason);
        const int length = strlen(test->todoReason);
        assert(length < DUMMY_MAX_MESSAGE_LENGTH);
        if(length > 0)
            return test->todoReason;
    }

    return NULL;
}

const char* dummyFormatV( const char* format, va_list args )
{
    static char buffer[DUMMY_MAX_MESSAGE_LENGTH];
    vsprintf(buffer, format, args);
    return buffer;
}

void dummyAbortTest( dummyTestAbortType type, const char* reason, ... )
{
    int errorCode = DUMMY_SANDBOX_GENERIC_ERROR;
    switch(type)
    {
        case DUMMY_FAIL_TEST:
            errorCode = DUMMY_SANDBOX_GENERIC_ERROR;
            break;

        case DUMMY_SKIP_TEST:
            errorCode = DUMMY_SANDBOX_SKIPPED;
            break;

        default:
            assert(!"Unknown abort type");
    }

    const char* formattedReason = NULL;
    if(reason)
    {
        va_list args;
        va_start(args, reason);
        formattedReason = dummyFormatV(reason, args);
        va_end(args);
    }
    else
    {
        assert(!"dummyAbortTest needs a reason!");
    }

    dummyAbortSandbox(errorCode, formattedReason);
}

void dummyMarkTestAsTodo( const char* reason, ... )
{
    dummyTest* test = dummyGetCurrentTest();
    test->markedAsTodo = true;
    if(reason)
    {
        va_list args;
        va_start(args, reason);
        strncpy(test->todoReason, dummyFormatV(reason, args), DUMMY_MAX_MESSAGE_LENGTH);
        va_end(args);
    }
}

void dummyLog( const char* message, ... )
{
    const dummyContext* ctx = dummyCurrentContext;
    assert(ctx);

    va_list args;
    va_start(args, message);
    const char* formattedMessage = dummyFormatV(message, args);
    va_end(args);

    ctx->reporter->log(ctx->reporter->context, formattedMessage);
}

void dummyPushAbortHandler( dummyAbortHandler handler, void* context )
{
    assert(AbortStackSize <= DUMMY_MAX_SANDBOX_DEPTH);
    AbortStackSize++;
    dummyAbortHandlerInfo* info = &AbortStack[AbortStackSize-1];

    info->handler = handler;
    info->context = context;
}

void dummyPopAbortHandler()
{
    assert(AbortStackSize >= 1);
    AbortStackSize--;
}

void dummyAbortSandbox( int errorCode, const char* reason )
{
    assert(AbortStackSize >= 1);
    dummyAbortHandlerInfo* info = &AbortStack[AbortStackSize-1];

    info->handler(info->context, errorCode, reason);
}
