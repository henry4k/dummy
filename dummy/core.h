#ifndef __DUMMY_CORE_H__
#define __DUMMY_CORE_H__

#ifdef __cplusplus
extern "C"
{
#endif

enum
{
    DUMMY_UNKNOWN_TEST_COUNT = -1
};

typedef enum
{
    DUMMY_TEST_PASSED,
    DUMMY_TEST_FAILED,
    DUMMY_TEST_SKIPPED
} dummyTestResult;

typedef enum
{
    DUMMY_FAIL_TEST,
    DUMMY_SKIP_TEST
} dummyTestAbortType;

enum
{
    DUMMY_SANDBOX_SUCEEDED = 0,
    DUMMY_SANDBOX_GENERIC_ERROR = 1
};

/**
 * Function that is called for a test.
 */
typedef void (*dummyTestFunction)();

typedef void (*dummyCleanupFunction)( void* data );

typedef struct
{
    void* context;

    /**
     * Runs the given function and catches errors.
     *
     * @param abortReason
     * If the protected call is aborted with a reason,
     * the string pointer is passed here.
     *
     * @return
     * Code that classifies the error.
     * If the function ras successfully it returns DUMMY_PROTECTED_CALL_SUCEEDED.
     * Custom error codes may be passed using dummySandbox::abort().
     */
    int (*run)( void* context, dummyTestFunction fn, const char** abortReason );

    /**
     * Aborts the current test.
     *
     * @param reason
     * May be `NULL`.
     *
     * @return
     * Doesn't return.
     */
    void (*abort)( void* context, int errorCode, const char* reason );
} dummySandbox;

typedef struct
{
    void* context;

    void (*began)( void* context );
    void (*completed)( void* context );

    void (*beganTest)( void* context );
    void (*completedTest)( void* context );

    void (*log)( void* context, const char* message );
} dummyReporter;


/**
 * Initializes the test context.
 *
 * This needs to be called before any calls to other test functions.
 * The initial context status is `DUMMY_INITIALIZING`.
 *
 * @param reporter
 * Reporter which will be used by the created context.
 */
void dummyInit( const dummySandbox* sandbox, const dummyReporter* reporter );

/**
 * Runs all added tests and destroys the current context.
 *
 * @return
 * Number of test tests, that failed.
 */
int dummyRunTests();

/**
 * Adds a test to the current context.
 *
 * @param name
 * May be used by the reporter to describe the test.
 *
 * @param fn
 * Is called when the test is being run.
 * See #dummyRunTests
 */
void dummyAddTest( const char* name, dummyTestFunction fn );

/**
 * Count of tests added to the current context.
 */
int dummyGetTestCount();

/**
 * Name of the active test in the current context.
 */
const char* dummyGetTestName();

/**
 * Id of the active test in the current context.
 */
int dummyGetTestNumber();

/**
 * TODO
 */
dummyTestResult dummyGetTestResult();

/**
 * Message describing why the current test aborted
 * or `NULL` if the test hasn't failed (yet).
 */
const char* dummyGetTestAbortReason();

/**
 * Adds a cleanup function to the current test.
 *
 * @param data
 * Data pointer that is passed to the cleanup function.
 */
void dummyAddCleanup( dummyCleanupFunction fn, void* data );

int dummyTestIsMarkedAsTodo();

/**
 * If set, it returns the reason why the current test has been marked as TODO.
 * May be `NULL` if no reason has been given.
 */
const char* dummyGetTestTodoReason();

/**
 * Aborts current test with the given reason.
 *
 * @param reason
 * Explains why the test was aborted -
 * this information is available to the reporter.
 *
 * @return
 * Doesn't return.
 */
void dummyAbortTest( dummyTestAbortType type, const char* reason, ... );

/**
 * Marks current test as TODO.
 *
 * @param reason
 * Explains why the test has been marked as TODO -
 * this information is available to the reporter.
 * May be `NULL` to signal an unknown reason.
 *
 * @note
 * Calling this function more than once in a test,
 * may overwrite the previous TODO reason.
 */
void dummyMarkTestAsTodo( const char* reason, ... );


/**
 * Reports a diagnostic message.
 */
void dummyLog( const char* message, ... );


#ifdef __cplusplus
}
#endif

#endif
