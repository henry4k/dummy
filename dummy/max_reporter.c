#include <stdio.h>
#include <assert.h>
#include "max_reporter.h"

typedef struct
{
    FILE* file;
} Context;

static void began( void* ctx_ )
{
    const Context* ctx = (Context*)ctx_;
    fprintf(ctx->file, "test count = %d\n",
        dummyGetTestCount()
    );
}

static void completed( void* ctx_ )
{
    const Context* ctx = (Context*)ctx_;
    fprintf(ctx->file, "testing done\n");
}

static void beganTest( void* ctx_ )
{
    const Context* ctx = (Context*)ctx_;
    fprintf(ctx->file, "test number = %d\n",
        dummyGetTestNumber()
    );
    fprintf(ctx->file, "test name = '%s'\n",
        dummyGetTestName()
    );
}

static void completedTest( void* ctx_ )
{
    const Context* ctx = (Context*)ctx_;

    const char* result = NULL;
    switch(dummyGetTestResult())
    {
        case DUMMY_TEST_PASSED:
            result = "passed";
            break;

        case DUMMY_TEST_FAILED:
            result = "failed";
            break;

        case DUMMY_TEST_SKIPPED:
            result = "skipped";
            break;

        default:
            assert(!"Unknown result");
    }
    fprintf(ctx->file, "test result = '%s'\n",
        result
    );
    fprintf(ctx->file, "test abort reason = '%s'\n",
        dummyGetTestAbortReason()
    );

    fprintf(ctx->file, "test todo = %d\n",
        dummyTestIsMarkedAsTodo()
    );
    fprintf(ctx->file, "test todo reason = '%s'\n",
        dummyGetTestTodoReason()
    );

    fprintf(ctx->file, "test done\n");
}

static void diag( void* ctx_, const char* message )
{
    const Context* ctx = (Context*)ctx_;
    fprintf(ctx->file, "%s\n",
        message
    );
}

const dummyReporter* dummyGetMaxReporter()
{
    static Context ctx;
    static dummyReporter reporter;

    ctx.file = stdout;

    reporter.context = &ctx;
    reporter.began = began;
    reporter.completed = completed;
    reporter.beganTest = beganTest;
    reporter.completedTest = completedTest;
    reporter.diag = diag;

    return &reporter;
}
