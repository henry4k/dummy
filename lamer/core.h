#ifndef __LAMER_CORE_H__
#define __LAMER_CORE_H__

#ifdef __cplusplus
extern "C"
{
#endif

enum
{
    LAMER_UNKNOWN_TEST_COUNT = -1
};

typedef enum
{
    LAMER_INITIALIZING,
    LAMER_RUNNING,
    LAMER_COMPLETED
} lamerStatus;

typedef enum
{
    LAMER_TEST_UNDEFINED,
    LAMER_TEST_STARTING,
    LAMER_TEST_RUNNING,
    LAMER_TEST_COMPLETED
} lamerTestStatus;

typedef enum
{
    LAMER_TEST_PASSED,
    LAMER_TEST_FAILED,
    LAMER_TEST_SKIPPED
} lamerTestResult;

typedef enum
{
    LAMER_FAIL_TEST,
    LAMER_SKIP_TEST
} lamerTestAbortType;

typedef struct
{
    void* context;

    void (*began)( void* context );
    void (*completed)( void* context );

    void (*beganTest)( void* context );
    void (*completedTest)( void* context );

    void (*diag)( void* context, const char* message );
} lamerReporter;


/**
 * Initializes the test context.
 *
 * This needs to be called before any calls to other test functions.
 * The initial context status is `LAMER_INITIALIZING`.
 *
 * @param reporter
 * Reporter which will be used by the created context.
 */
void lamerInit( const lamerReporter* reporter );

/**
 * Runs all added tests and destroys the current context.
 *
 * @return
 * Number of test tests, that failed.
 */
int lamerRunTests();

/**
 * Function that is called for a test.
 */
typedef void (*lamerTestFunction)();

/**
 * Adds a test to the current context.
 *
 * @param name
 * May be used by the reporter to describe the test.
 *
 * @param fn
 * Is called when the test is being run.
 * See #lamerRunTests
 *
 * @return
 * Id of the test.
 */
int lamerAddTest( const char* name, lamerTestFunction fn );

/**
 * Status of the current context.
 */
lamerStatus lamerGetStatus();

/**
 * Count of tests added to the current context.
 */
int lamerGetTestCount();

/**
 * Name of the active test in the current context.
 */
const char* lamerGetTestName();

/**
 * Id of the active test in the current context.
 */
int lamerGetTestNumber();

/**
 * Status of the active test in the current context.
 */
lamerTestStatus lamerGetTestStatus();

/**
 * TODO
 */
lamerTestResult lamerGetTestResult();

/**
 * Message describing why the current test aborted
 * or `NULL` if the test hasn't failed (yet).
 */
const char* lamerGetTestAbortReason();

int lamerTestIsMarkedAsTodo();

const char* lamerGetTestTodoReason();

/**
 * Aborts current test with the given reason.
 *
 * @param reason
 * Explains why the test was aborted - this information is available
 * to the reporter. May be `NULL` to signal an unknown reason.
 *
 * @return
 * Doesn't return.
 */
void lamerAbortTest( lamerTestAbortType type, const char* reason, ... );

/**
 * Marks current test as TODO.
 *
 * @param reason
 * Explains why the test has been marked as TODO -
 * this information is available to the reporter.
 * May be `NULL` to signal an unknown reason.
 *
 * @note
 * Calling this function more than once in a test,
 * may overwrite the previous TODO reason.
 */
void lamerMarkTestAsTodo( const char* reason, ... );


#ifdef __cplusplus
}
#endif

#endif
