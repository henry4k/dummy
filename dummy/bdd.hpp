#ifndef __LAMER_BDD_HPP__
#define __LAMER_BDD_HPP__

#include <string>
#include "core.h"

namespace lamer
{

class Describe
{
public:
    Describe( const char* subject );
    Describe& it( const char* behaviour, lamerTestFunction fn );

private:
    std::string m_Subject;
};


// ---- implementation ----

Describe::Describe( const char* subject ) :
    m_Subject(subject)
{
}

Describe& Describe::it( const char* behaviour, lamerTestFunction fn )
{
    lamerAddTest((m_Subject+" "+behaviour).c_str(), fn);
    return *this;
}

}

#endif
