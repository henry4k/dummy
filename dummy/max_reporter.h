#ifndef __DUMMY_MAX_REPORTER_H__
#define __DUMMY_MAX_REPORTER_H__

#include "core.h"

#ifdef __cplusplus
extern "C"
{
#endif

/**
 * Reports as many information as possible using a custom format.
 *
 * @note
 * It is used for self dummying and not intended to be used externally,
 * since the format may change over time.
 */
const dummyReporter* dummyGetMaxReporter();

#ifdef __cplusplus
}
#endif

#endif
