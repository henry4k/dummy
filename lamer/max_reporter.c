#include <stdio.h>
#include <assert.h>
#include "max_reporter.h"

typedef struct
{
    FILE* file;
} Context;

static void began( void* ctx_ )
{
    const Context* ctx = (Context*)ctx_;
    fprintf(ctx->file, "test count = %d\n",
        lamerGetTestCount()
    );
}

static void completed( void* ctx_ )
{
    const Context* ctx = (Context*)ctx_;
    fprintf(ctx->file, "testing done\n");
}

static void beganTest( void* ctx_ )
{
    const Context* ctx = (Context*)ctx_;
    fprintf(ctx->file, "test number = %d\n",
        lamerGetTestNumber()
    );
    fprintf(ctx->file, "test name = '%s'\n",
        lamerGetTestName()
    );
}

static void completedTest( void* ctx_ )
{
    const Context* ctx = (Context*)ctx_;

    const char* result = NULL;
    switch(lamerGetTestResult())
    {
        case LAMER_TEST_PASSED:
            result = "passed";
            break;

        case LAMER_TEST_FAILED:
            result = "failed";
            break;

        case LAMER_TEST_SKIPPED:
            result = "skipped";
            break;

        default:
            assert(!"Unknown result");
    }
    fprintf(ctx->file, "test result = '%s'\n",
        result
    );
    fprintf(ctx->file, "test abort reason = '%s'\n",
        lamerGetTestAbortReason()
    );

    fprintf(ctx->file, "test todo = %d\n",
        lamerTestIsMarkedAsTodo()
    );
    fprintf(ctx->file, "test todo reason = '%s'\n",
        lamerGetTestTodoReason()
    );

    fprintf(ctx->file, "test done\n");
}

static void diag( void* ctx_, const char* message )
{
    const Context* ctx = (Context*)ctx_;
    fprintf(ctx->file, "%s\n",
        message
    );
}

const lamerReporter* lamerGetMaxReporter()
{
    static Context ctx;
    static lamerReporter reporter;

    ctx.file = stdout;

    reporter.context = &ctx;
    reporter.began = began;
    reporter.completed = completed;
    reporter.beganTest = beganTest;
    reporter.completedTest = completedTest;
    reporter.diag = diag;

    return &reporter;
}
