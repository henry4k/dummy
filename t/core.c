#include <stdio.h> // setvbuf
#include <stddef.h> // NULL
#include <assert.h>
#include "../test/core.h"
#include "../test/max_reporter.h"

void PassingTest()
{
}

void AssertTest()
{
    assert(0);
}

void FailTest()
{
    testFailCase(NULL);
}

void FailWithReasonTest()
{
    testFailCase("Failed with a reason");
}

void SkipTest()
{
    testSkipCase(NULL);
}

void SkipWithReasonTest()
{
    testSkipCase("Skipped with a reason");
}

void TodoTest()
{
    testMarkCaseAsTodo(NULL);
}

void TodoWithReasonTest()
{
    testMarkCaseAsTodo("TODO with a reason");
}

int main()
{
    // Disable buffering, so stdout and stderr are merged correctly
    setvbuf(stdout, NULL, _IONBF, 0);
    setvbuf(stderr, NULL, _IONBF, 0);

    testInit(testGetMaxReporter());
    testAddCase("PassingTest", PassingTest);
    testAddCase("AssertTest", AssertTest);
    testAddCase("FailTest", FailTest);
    testAddCase("FailWithReasonTest", FailWithReasonTest);
    testAddCase("SkipTest", SkipTest);
    testAddCase("SkipWithReasonTest", SkipWithReasonTest);
    testAddCase("TodoTest", TodoTest);
    testAddCase("TodoWithReasonTest", TodoWithReasonTest);
    testRun();
    return 0;
}
