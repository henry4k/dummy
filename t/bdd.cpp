#include "../test/core.h"
#include "../test/max_reporter.h"
#include "../test/bdd.hpp"

using namespace test;

int main()
{
    testInit(testGetMaxReporter());

    Describe("fizzbuzzer module")
        .it("does things", [](){
            // try doing things here
        })
        .it("is a slacker", [](){
            testMarkCaseAsTodo("Implement lazyness");
            // prove that fizzbuzzer is lazy as fuck
        })
        .it("can't handle bloops", [](){
            testMarkCaseAsTodo("Implement exception");
            // test if bloops cause fizzbuzzer to fail
            testSkipCase("Exception not implemented yet");
        });

    Describe("bloop")
        .it("can make a sound", [](){
            // emulate a sound
            testFailCase("Bloop is silent!");
        })
        .it("looks weird", [](){
            testMarkCaseAsTodo("Bloop needs to look weirder.");
            // bloops should always look weird
            testFailCase("Bloop doesn't looks weird enough!");
        });

    testRun();
}
