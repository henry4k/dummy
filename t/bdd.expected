test case count = 5
test case number = 0
test case name = 'fizzbuzzer module does things'
test case result = 'passed'
test case abort reason = '(null)'
test case todo = 0
test case todo reason = '(null)'
test case done
test case number = 1
test case name = 'fizzbuzzer module is a slacker'
test case result = 'passed'
test case abort reason = '(null)'
test case todo = 1
test case todo reason = 'Implement lazyness'
test case done
test case number = 2
test case name = 'fizzbuzzer module can't handle bloops'
test case result = 'skipped'
test case abort reason = 'Exception not implemented yet'
test case todo = 1
test case todo reason = 'Implement exception'
test case done
test case number = 3
test case name = 'bloop can make a sound'
test case result = 'failed'
test case abort reason = 'Bloop is silent!'
test case todo = 0
test case todo reason = '(null)'
test case done
test case number = 4
test case name = 'bloop looks weird'
test case result = 'failed'
test case abort reason = 'Bloop doesn't looks weird enough!'
test case todo = 1
test case todo reason = 'Bloop needs to look weirder.'
test case done
testing done
